app/src/main/kotlin/org/kotlinlsp/common/Log.kt

```diff
@@ -29,21 +33,61 @@ private enum class LogLevel(level: Int) {
 private val logLevel = LogLevel.Debug
 private const val profileEnabled = true
 
-private lateinit var logFile: File
+private lateinit var logger: LSPLogger
 private val profileInfo = mutableMapOf<String, Pair<Int, Duration>>()
 
-fun setupLogger(path: String) {
-    val loggerPath = "$path/log.txt"
-    logFile = File(loggerPath)
-    if (logFile.exists()) {
-        logFile.delete()
+private class LSPLogger {
+    var client: LanguageClient? = null
+    val messageQueue = ConcurrentLinkedQueue<MessageParams>()
+
+    fun log(level: LogLevel, message: String) {
+        if (level < logLevel) return
+
+        val type = when (level) {
+            LogLevel.Trace -> MessageType.Log
+            LogLevel.Debug -> MessageType.Log
+            LogLevel.Info -> MessageType.Info
+            LogLevel.Warning -> MessageType.Warning
+            LogLevel.Error -> MessageType.Error
+            LogLevel.Off -> return
+        }
+
+        client?.logMessage(MessageParams(type, message))
+            ?: messageQueue.add(MessageParams(type, message))
     }
 
+    fun redirectSystemErr() {
+        val loggerStream = PrintStream(object : OutputStream() {
+            private val buffer = StringBuilder()
+
+            override fun write(b: Int) {
+                if (b == '\n'.code) {
+                    log(LogLevel.Error, buffer.toString())
+                    buffer.setLength(0)
+                } else {
+                    buffer.append(b.toChar())
+                }
+            }
+        })
+
+        System.setErr(loggerStream)
+    }
+}
+
+fun setupLogger() {
```

Comment:
Right now it is crashing at startup, in my case `connect` is being called before `initialized` so `logger` is not initialized. I think we can just call `setupLogger` in `initialize` to avoid this issue

Reply:
Curious, I tried to call it at initialize, but it is called before connect for me?

Reply:
Oh, it is called before initialize in the tests

Reply:
Whoops, I should fix that as well

Reply:
I logged the calls and this is the order for me:
- Connect
- Initialize
- Initialized